storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +63,403 @@
-                    Long offset = partitionToConsumedOffsets.get(metadataPartition);
-                    if (offset != null) {
-                        remotePartitionMetadataEventHandler.syncLogMetadataSnapshot(topicIdPartition, metadataPartition, offset);
+        maybeFetchStartAndEndOffsets();
+        boolean isAllInitialized = true;
+        for (final UserTopicIdPartition utp : assignedUserTopicIdPartitions.values()) {
+            if (utp.isAssigned && !utp.isInitialized) {
+                final Integer metadataPartition = utp.metadataPartition;
+                final StartAndEndOffsetHolder holder = offsetHolderByMetadataPartition.get(toRemoteLogPartition(metadataPartition));
+                // The offset-holder can be null, when the recent assignment wasn't picked up by the consumer.
+                if (holder != null) {
+                    final Long readOffset = readOffsetsByMetadataPartition.getOrDefault(metadataPartition, -1L);
+                    // 1) The end-offset was fetched only once during reassignment. The metadata-partition can receive
+                    // new stream of records, so the consumer can read records more than the last-fetched end-offset.
+                    // 2) When the internal topic becomes empty due to breach by size/time/start-offset, then there
+                    // are no records to read.
+                    if (readOffset + 1 >= holder.endOffset || holder.endOffset.equals(holder.startOffset)) {
+                        markInitialized(utp);
                     } else {
-                        log.debug("Skipping sync-up of the remote-log-metadata-file for partition: [{}] , with remote log metadata partition{}, and no offset",
-                                topicIdPartition, metadataPartition);
+                        log.debug("The user-topic-partition {} could not be marked initialized since the read-offset is {} " +
+                                "but the end-offset is {} for the metadata-partition {}", utp, readOffset, holder.endOffset,
+                            metadataPartition);
                     }
+                } else {
+                    log.debug("The offset-holder is null for the metadata-partition {}. The consumer may not have picked" +
+                            " up the recent assignment", metadataPartition);
                 }
-
-                // Write partitionToConsumedOffsets into committed offsets file as we do not want to process them again
-                // in case of restarts.
-                committedOffsetsFile.writeEntries(partitionToConsumedOffsets);
-                lastSyncedPartitionToConsumedOffsets = new HashMap<>(partitionToConsumedOffsets);
             }
-
-            lastSyncedTimeMs = time.milliseconds();
-        } catch (IOException e) {
-            throw new KafkaException("Error encountered while writing committed offsets to a local file", e);
+            isAllInitialized = isAllInitialized && utp.isInitialized;
```

Comment:
That should be ok. @kamalcph do you have any concerns? Since this is single-threaded, I don't see a race happening. The function where we mark utp as assigned (`maybeWaitForPartitionAssignments`) and this method are executed by the same thread.